Support for acquisition board v3

aacuevas opened the pull request:
Adds support for new acquisition boards, that allow the use of BNO-capable headstages. Most changes are to the AcqBoardONI and related classes, with some changes on the editor when using this particular board.

- Fixes #6 
- Fixes #18 
- Fixes #21 
- Fixes #22 
- Fixes #23 
- Fixes #24 
- Fixes #29 
- Fixes #30 
- Fixes #33

aacuevas committed:
Add BNO and memory stream support

aacuevas committed:
Fix version returns on ONI-version of API

aacuevas committed:
Merge remote-tracking branch 'origin/main' into acqboard-v3-support

bparks13 committed:
Add support for discovering connected BNOs across all ports

bparks13 committed:
Corrected firmware version check

bparks13 committed:
Minor updates
- Only check firmware version once while opening the board
- Consolidate enable/disable methods into a single method that takes an array and enables/disables as needed

bparks13 committed:
Add support for BNO streams on all ports

bparks13 committed:
Create array of DataBuffer pointers to handle data streams

aacuevas committed:
Fixes to BNO stream generation

aacuevas committed:
Merge remote-tracking branch 'origin/main' into acqboard-v3-support

bparks13 committed:
Address firmware changes based on #18
- Updates the BNO device indices and register locations
- Adds I2C raw device indices and registers
- During discovery, check if there is an I2C capable device on each port
- Read the hardware ID from the EEPROM if an I2C capable device exists
- Set the axis map for the BNO if one exists

bparks13 committed:
Find device address for methods based on the given port number
- Reorder parameters to be consistent with oni_* calls
- Before reading device IDs from an EEPROM, confirm that the first four bytes contain the characters "OESH"
- Make readBytes() more flexible, and create readEepromBytes() specifically for EEPROM byte reads
- Corrected logic for devices with a BNO but no EEPROM

bparks13 committed:
Addressed review comments
- Modified Rhd2000ONIBoard::readByte to be more generalized and flexible
- Expanded the EEPROM identifier to include "\x01", which is the spec version
- Corrected a bug that was accessing the wrong device index when checking if a BNO is connected

bparks13 committed:
Return patch version when getting firmware version
- Check what version is returned, and confirm if it is I2C capable

bparks13 committed:
Display "IMU" when BNO is connected to a port

bparks13 committed:
Check if board is ready to stream

bparks13 committed:
Get the acquisition clock rate

bparks13 committed:
Set memory monitor sample rate
- Sets the sample rate to be 100 Hz
- Memory monitor is now the percent of total memory used

bparks13 committed:
Allow streaming for BNO-only headstages
- Refactor how buffers are created and resized across all devices
- Ensure that the editor is correctly updated if a BNO moves

bparks13 committed:
Update unmanaged registers address

bparks13 committed:
Only show ephys channels in canvas

bparks13 committed:
Add MemoryMonitor usage indicator to editor
- Add a `BoardType` member so that different boards can be easily identified
- Shift the editor over if the board is of type `ONI`, so there is room for the memory monitor level component
- Add error reporting to memory monitor read/writes

bparks13 committed:
Add timeout for checking if BNO is connected

bparks13 committed:
Detect board version
- Add method to check the device ID when opening the board
- For V3, modify how the ADC/DAC values are handled

bparks13 committed:
Merge pull request #19 from open-ephys-plugins/issue-18

Address initial comments on updates to firmware

bparks13 committed:
I2C support begins in version 1.5.0
- Update LOG output to include all version components

bparks13 committed:
Merge pull request #25 from open-ephys-plugins/issue-6

Add BNO functionality

bparks13 committed:
Merge pull request #27 from open-ephys-plugins/issue-23

Detect board version

bparks13 committed:
Initialize member
- Set isTransmitting to false during construction

bparks13 committed:
Ensure EEPROM headstage ID is read correctly

aacuevas committed:
Update liboni dlls

aacuevas committed:
Merge pull request #26 from open-ephys-plugins/issue-22

Add MemoryMonitor usage indicator to editor

aacuevas committed:
Add prerelease info to gateware version

bparks13 committed:
Show IMU on right slot only

aacuevas committed:
Merge pull request #28 from open-ephys-plugins/v3-smallfixes

Some extra fixes

aacuevas committed:
Read EEPROM id in v1 and v2 of the format

bparks13 committed:
Update stream and channel identifiers

bparks13 committed:
Add all BNO055 data channels to the stream

aacuevas committed:
Merge pull request #31 from open-ephys-plugins/eeprom-version

Read EEPROM id in v1 and v2 of the format

bparks13 committed:
Give bitVolts and units to channels

bparks13 committed:
Add information to memory identifier, fix Euler labels

bparks13 committed:
Merge pull request #32 from open-ephys-plugins/issue-30

Add BNO channels to stream, and update stream/channel identifiers

bparks13 committed:
Separate calibration status into four distinct channels

aacuevas committed:
Merge pull request #34 from open-ephys-plugins/issue-33

Separate calibration status into four distinct channels

bparks13 commented:
@anjaldoshi I think the only thing missing from this PR is statically linking the liboni library instead of dynamically linking it. Can you provide any insight into the best way to implement this using cmake? We should implement this library change before we package the plugin and upload it.

jsiegle commented:
I'm testing this out now...I'm not seeing any data coming in on the memory usage stream (using OE FPGA acq board). Even if it were working, I'm not sure whether this is worth including as its own continuous data stream. What if instead the plugin sent a broadcast message every time the memory usage increased or decreased by 10%, e.g. `Acquisition Board (102) memory usage = 20%"`...that way there will be a record of memory usage issues in the recorded data, but it won't require a separate data stream.

bparks13 committed:
Always enable memory monitor

bparks13 committed:
Enable memory monitor when detecting the board

jsiegle commented:
@aacuevas can you point me to the latest firmware for pre-v3 boards? It sounds like I will need that for the memory monitor to work properly

aacuevas commented:
The memory monitor is in all versions, but there was a bug that made it look wrong. 
@bparks13 maybe we should disable the memory monitor if fw version < 1.5 ?

All boards, pre-v3 and v3 alike, can use the latest 1.5.1 firmware, we made it so it works across all boards with our FPGA, and should be the recommended fw (it does have some bugfixes).

you can find it [here](https://github.com/open-ephys/acq-board-docs/tree/acq-board-3d/source/_static/files) for the time being, until that page goes live

bparks13 committed:
Correctly check duration length
- For BNO devices, the wrong order of variables was used to check if a certain amount of time had passed

aacuevas committed:
Add BNO mappings for known headstages

bparks13 committed:
Only enable memory monitor for firmware >= v1.5.1

aacuevas committed:
Merge pull request #36 from open-ephys-plugins/check-memory-monitor-support

Only enable memory monitor for firmware >= v1.5.1

bparks13 committed:
Merge branch 'main' into acqboard-v3-support

bparks13 committed:
Update linux runner to use Ubuntu 22.04

jsiegle reviewed (approved):
Tested with OE FPGA and Opal Kelly acq boards, and everything looks good!
